visualization/app/codeCharta/util/algorithm/streetLayout/verticalStreet.ts

```diff
@@ -0,0 +1,188 @@
+		return origin.x + this.spacer + maxLeftWidth
+	}
+
+	private calculateChildOriginY(origin: Vector2, index: number, row: BoundingBox[]): number {
+		return origin.y + this.getLengthUntil(row, index)
+	}
+
+	private getLength(boxes: BoundingBox[]): number {
+		return this.getLengthUntil(boxes, boxes.length)
+	}
+
+	private getLengthUntil(boxes: BoundingBox[], end: number): number {
+		let sum = 0
+
+		for (let index = 0; index < end; index++) {
+			sum += boxes[index].height
+		}
+		return sum
+	}
+
+	protected splitChildrenToRows(children: BoundingBox[]) {
+		const totalLength = this.getLength(children)
+		let sum = 0
+
+		for ( const child of children ) {
+			if (sum < totalLength / 2) {
+				if (child instanceof HorizontalStreet) {
+					(<HorizontalStreet>child).orientation = HorizontalOrientation.LEFT
+				}
+				this.leftRow.push(child)
+				sum += child.height
+			} else {
+				this.rightRow.push(child)
+			}
+		}
+	}
+
+	protected rearrangeRows(): void {
+		if (this.orientation === VerticalOrientation.UP) {
+			this.leftRow = [...this.leftRow].reverse()
+		} else {
+			this.rightRow = [...this.rightRow].reverse()
+		}
+	}
+
+	private getMaxWidth(boxes: BoundingBox[]): number {
+		return boxes.reduce((max, n) => Math.max(max, n.width), Number.MIN_VALUE)
+	}
+
+	protected calculateStreetOverhang(streetOrigin: Vector2): number {
+		if (this.orientation === VerticalOrientation.UP) {
+			return this.calculateTopStreetOverhang(streetOrigin)
+		}
+		return this.calculateBottomStreetOverhang(streetOrigin)
+	}
+
+	private calculateTopStreetOverhang(streetOrigin: Vector2): number {
+		const firstLeftNode = this.leftRow[0]
+		const firstRightNode = this.rightRow[0]
+		const leftOverhang =
+			firstLeftNode instanceof HorizontalStreet && firstLeftNode.streetRect
+				? firstLeftNode.streetRect.topLeft.y - streetOrigin.y
+				: this.height - this.getLength(this.leftRow)
+		const rightOverhang =
+			firstRightNode instanceof HorizontalStreet && firstRightNode.streetRect
+				? firstRightNode.streetRect.topLeft.y - streetOrigin.y
+				: this.height - this.getLength(this.rightRow)
+
+		return leftOverhang > 0 && rightOverhang > 0 ? Math.min(leftOverhang, rightOverhang) : 0
+	}
+
+	private calculateBottomStreetOverhang(streetOrigin: Vector2): number {
+		const lastLeftNode = this.leftRow[this.leftRow.length - 1]
+		const lastRightNode = this.rightRow[this.rightRow.length - 1]
+		const streetBottomY = streetOrigin.y + this.height
+		const leftOverhang =
+			lastLeftNode instanceof HorizontalStreet && lastLeftNode.streetRect
+				? streetBottomY - lastLeftNode.streetRect.getBottomRight().y
+				: this.height - this.getLength(this.leftRow)
+		const rightOverhang =
+			lastRightNode instanceof HorizontalStreet && lastRightNode.streetRect
+				? streetBottomY - lastRightNode.streetRect.getBottomRight().y
+				: this.height - this.getLength(this.rightRow)
+
+		return leftOverhang > 0 && rightOverhang > 0 ? Math.min(leftOverhang, rightOverhang) : 0
+	}
+}
```

Comment:
The code has a lot of code that looks very similar. I did not yet check if it's possible to abstract it but we might want to have a closer look at that :)